src/System.Runtime.Numerics/src/System/Numerics/BigIntegerCalculator.DivRem.cs

```diff
@@ -128,203 +129,194 @@ public unsafe static uint[] Remainder(uint[] left, uint[] right)
-                    target[i] = (value[i] << shift)
-                        | (value[i - 1] >> backShift);
-                }
-                if (targetLength > valueLength)
-                {
-                    target[valueLength] =
-                        value[valueLength - 1] >> backShift;
-                }
-            }
-            else
+            // Repairs the dividend, if the last subtract was too much
+
+            int i = 0;
+            ulong carry = 0L;
+
+            for (; i < rightLength; i++)
             {
-                for (int i = 0; i < valueLength; i++)
-                    target[i] = value[i];
+                ulong digit = (left[i] + carry) + right[i];
+                left[i] = (uint)digit;
+                carry = digit >> 32;
             }
+
+            return (uint)carry;
         }
 
         [SecuritySafeCritical]
-        private unsafe static void RightShift(uint* value, int valueLength,
-                                              uint* target, int targetLength,
-                                              int shift)
+        private unsafe static uint SubtractDivisor(uint* left, int leftLength,
+                                                   uint* right, int rightLength,
+                                                   ulong q)
         {
-            Debug.Assert(valueLength >= 1);
-            Debug.Assert(targetLength == valueLength);
-            Debug.Assert(shift >= 0 && shift < 32);
+            Debug.Assert(leftLength >= 0);
+            Debug.Assert(rightLength >= 0);
+            Debug.Assert(leftLength >= rightLength);
+            Debug.Assert(q <= 0xFFFFFFFF);
 
-            if (shift > 0)
-            {
-                int backShift = 32 - shift;
-                for (int i = 0; i < valueLength - 1; i++)
-                {
-                    target[i] = (value[i] >> shift)
-                        | (value[i + 1] << backShift);
-                }
-                target[valueLength - 1] =
-                    (value[valueLength - 1] >> shift);
-            }
-            else
+            // Combines a subtract and a multiply operation, which is naturally
+            // more efficient than multiplying and then subtracting...
+
+            int i = 0;
+            ulong carry = 0L;
+
+            for (; i < rightLength; i++)
             {
-                for (int i = 0; i < valueLength; i++)
-                    target[i] = value[i];
+                carry += right[i] * q;
+                uint digit = (uint)carry;
+                carry = carry >> 32;
+                if (left[i] < digit)
+                    ++carry;
+                left[i] -= digit;
             }
+
+            return (uint)carry;
         }
 
-        [SecuritySafeCritical]
-        private unsafe static void MultiplyDivisor(uint* left, int leftLength,
-                                                   ulong right, uint* bits)
+        private static bool CompareDivisor(uint divHi, uint divLo,
```

Comment:
Yes, but that requires someone to read the implementation to understand the signature of a method.